origin-discovery/src/listener/rules.js

```diff
@@ -20,6 +20,28 @@ function generateOfferId(log) {
   ].join('-')
 }
 
+/**
+ * Ensures data fetched from the blockchain meets the freshness criteria
+ * specified in blockInfo. This is to catch the case where data is fetched from
+ * an out of sync node that returns stale data.
+ * @param {List(Event)} events
+ * @param {blockNumber: number, logIndex: number} blockInfo
+ * @throws {Error} If freshness check fails
+ */
+function checkFreshness(events, blockInfo) {
+  // Loop thru events to ensure at least one of them is as fresh as blockInfo.
+  let fresh = false
+  events.forEach(event => {
+    if ((event.blockNumber > blockInfo.blockNumber) ||
+      (event.blockNumber === blockInfo.blockNumber && event.logIndex >= blockInfo.logIndex)) {
+      fresh = true
+    }
+  })
+  if (!fresh) {
+    throw new Error('Freshness check failed')
+  }
```

Comment:
a shorter alternative: 
```
// Find at least 1 event that is as fresh as blockInfo.
if (!events.some(event => {
  return (event.blockNumber > blockInfo.blockNumber) ||
    (event.blockNumber === blockInfo.blockNumber && event.logIndex >= blockInfo.logIndex)
})) {
  throw new Error('Freshness check failed')
}
```

Reply:
oh I did not know about some method on arrays. very useful. implemented. thank you ! :)